projects/RabbitMQ.Client/client/api/ConnectionFactory.cs

```diff
@@ -539,10 +659,52 @@ public IConnection CreateConnection(IEndpointResolver endpointResolver, string c
             {
                 if (AutomaticRecoveryEnabled)
                 {
-                    return new AutorecoveringConnection(config, endpointResolver);
+                    var c = new AutorecoveringConnection(config, endpointResolver);
+                    return (AutorecoveringConnection)c.Open();
+                }
+                else
+                {
+                    var c = new Connection(config, endpointResolver.SelectOne(CreateFrameHandler));
+                    return (Connection)c.Open();
                 }
+            }
+            catch (Exception e)
+            {
+                throw new BrokerUnreachableException(e);
+            }
+        }
 
-                return new Connection(config, endpointResolver.SelectOne(CreateFrameHandler));
+        /// <summary>
+        /// Asynchronously create a connection using an IEndpointResolver.
+        /// </summary>
+        /// <param name="endpointResolver">
+        /// The endpointResolver that returns the endpoints to use for the connection attempt.
+        /// </param>
+        /// <param name="clientProvidedName">
+        /// Application-specific connection name, will be displayed in the management UI
+        /// if RabbitMQ server supports it. This value doesn't have to be unique and cannot
+        /// be used as a connection identifier, e.g. in HTTP API requests.
+        /// This value is supposed to be human-readable.
+        /// </param>
+        /// <returns>Open connection</returns>
+        /// <exception cref="BrokerUnreachableException">
+        /// When no hostname was reachable.
+        /// </exception>
+        public ValueTask<IConnection> CreateConnectionAsync(IEndpointResolver endpointResolver, string clientProvidedName)
+        {
+            ConnectionConfig config = CreateConfig(clientProvidedName);
+            try
+            {
+                if (AutomaticRecoveryEnabled)
+                {
+                    var c = new AutorecoveringConnection(config, endpointResolver);
+                    return c.OpenAsync();
```

Comment:
This is not going to work if you intent the catch block to capture exceptions from openasync. You'd have to await here

Reply:
Thank you for noticing that. I will make a note to review my changes for this situation.